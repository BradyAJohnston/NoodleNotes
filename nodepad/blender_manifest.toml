schema_version = "1.0.0"

id = "molecularnodes"
version = "0.0.9"
name = "NodePad"
tagline = "Create beautiful documentation for your node groups"
maintainer = "Brady Johnston<brady.johnston@me.com>"
type = "add-on"
website = "https://bradyajohnston.github.io/nodepad"
tags = [
	"Geometry Nodes"
]
blender_version_min = "4.2.0"
license = [
  "SPDX:GPL-3.0-or-later"
]
copyright =[
	"2024 Brady Johnston",
]