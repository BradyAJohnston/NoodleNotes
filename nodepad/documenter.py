import bpy
import pathlib
import sys

from .interface import InterfaceGroup, InterfaceItem
from . import markdown
from typing import List

TOP_FOLDER = pathlib.Path(__file__).resolve().parent.parent.parent
sys.path.insert(0, str(TOP_FOLDER))


class Title(str):
    def as_markdown(self, level: int = 2) -> str:
        return "{} {}".format("#" * level, self)


class Video(str):
    def as_markdown(self, text="") -> str:
        string = self
        if not string.endswith(".mp4"):
            string += ".mp4"
        return f"![{text}]({string})\n"


class Description(str):
    def as_markdown(self) -> str:
        if self == "":
            return ""
        return "\n{}\n".format(self)


class Videos(list):
    def as_markdown(self) -> str:
        return "\n".join([x.as_markdown() for x in self])


class Documenter:
    def __init__(self, tree: bpy.types.NodeTree) -> None:
        self.tree = tree
        self.level: int = 2
        self.title: Title = Title(tree.name)
        self.items = [InterfaceItem(x) for x in tree.interface.items_tree]
        self.inputs = InterfaceGroup([x for x in self.items if x.is_input])
        self.outputs = InterfaceGroup(
            [x for x in self.items if x.is_output], is_output=True
        )
        self.level = 2
        self._description: Description = Description(tree.description)
        self._links: list[str] = []
        self._video_links: list[Video] = []

    @property
    def name(self) -> str:
        return self.tree.name

    @property
    def description(self) -> Description:
        return self._description

    @property
    def videos(self) -> Videos:
        return Videos([x for x in self._video_links])

    def lookup_info(self, extra_json: dict) -> None:
        try:
            for link in extra_json[self.name]["videos"]:
                self._video_links.append(Video(link))
        except KeyError:
            pass

    def collect_items(self):
        items = [
            self.title.as_markdown(level=self.level),
            self.description.as_markdown(),
            self.videos.as_markdown(),
            self.outputs.as_markdown("Outputs"),
            self.inputs.as_markdown("Inputs"),
        ]
        return [item for item in items if item is not None]

    def as_markdown(self) -> str:
        text = "\n"
        text += "\n".join(self.collect_items())
        return text


class TreeDocumenter(Documenter):
    def __init__(self, tree: bpy.types.NodeTree) -> None:
        super().__init__(tree=tree)
