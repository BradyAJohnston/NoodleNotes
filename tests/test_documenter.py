import bpy
import nodepad
from nodepad import Documenter
from pathlib import Path
import pytest

nodepad.register()

nodes = [
    "Blend Hair Curves",
    "Displace Hair Curves",
    "Frizz Hair Curves",
    "Roll Hair Curves",
    "Braid Hair Curves",
    "Curve Info",
    "Curve Root",
    "Attach Hair Curves to Surface",
]

DATADIR = Path(bpy.utils.script_paths()[0]).parent / "datafiles/assets/geometry_nodes/"


def test_documenter(snapshot):
    for node_name in nodes:
        bpy.ops.wm.append(
            "EXEC_DEFAULT",
            directory=str(DATADIR / "procedural_hair_node_assets.blend/NodeTree/"),
            filename=node_name,
            use_recursive=True,
        )

        assert snapshot == Documenter(bpy.data.node_groups[node_name]).as_markdown()


nodepad.unregister()
