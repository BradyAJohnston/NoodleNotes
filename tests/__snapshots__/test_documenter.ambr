# serializer version: 1
# name: test_documenter
  '''
  
  ## Blend Hair Curves
  
  
  
  ### Outputs
  
  |Description|Socket            |
  |-----------|------------------:|
  |           |`Geometry::Geometry`|
  
  : {tbl-colwidths="[90, 10]"}
  
  
  
  
  ### Inputs
  
  |Socket               |Default|Description                                    |
  |---------------------|:-------:|-----------------------------------------------|
  |`Geometry::Geometry` |_None_ |Input Geometry (may include other than curves) |
  |`Factor::Float`      |1.0    |Factor to blend overall effect                 |
  |`Blend Radius::Float`|0.05   |Radius to select neighbors for blending        |
  |`Blend Neighbors::Int`|10     |Amount of neighbors used for blending          |
  |`Preserve Length::Bool`|False  |Preserve each curve's length during deformation|
  
  : {tbl-colwidths="[10, 15, 80]"}
  
  
  
  '''
# ---
# name: test_documenter.1
  '''
  
  ## Curve Info
  
  
  
  ### Outputs
  
  |Description                                    |Socket            |
  |-----------------------------------------------|------------------:|
  |Index of each Curve                            |`Curve Index::Int`|
  |ID of each curve                               |`Curve ID::Int`   |
  |Length of each curve                           |`Length::Float`   |
  |Direction from root to tip of each curve       |`Direction::Vector`|
  |Random vector for each curve                   |`Random::Float`   |
  |Attachment surface UV coordinates of each curve|`Surface UV::Vector`|
  
  : {tbl-colwidths="[90, 10]"}
  
  
  
  
  
  '''
# ---
# name: test_documenter.2
  '''
  
  ## Curve Root
  
  
  
  ### Outputs
  
  |Description                             |Socket                |
  |----------------------------------------|----------------------:|
  |Boolean selection of curve root points  |`Root Selection::Bool`|
  |Position of the root point of a curve   |`Root Position::Vector`|
  |Direction of the root segment of a curve|`Root Direction::Vector`|
  |Index of the root point of a curve      |`Root Index::Int`     |
  
  : {tbl-colwidths="[90, 10]"}
  
  
  
  
  
  '''
# ---
# name: test_documenter.3
  '''
  
  ## Attach Hair Curves to Surface
  
  
  
  ### Outputs
  
  |Description                                   |Socket                       |
  |----------------------------------------------|-----------------------------:|
  |                                              |`Geometry::Geometry`         |
  |Surface UV coordinates at the attachment point|`Surface UV Coordinate::Vector`|
  |Surface normal at the attachment point        |`Surface Normal::Vector`     |
  
  : {tbl-colwidths="[90, 10]"}
  
  
  
  
  ### Inputs
  
  |Socket                       |Default        |Description                                                       |
  |-----------------------------|:---------------:|------------------------------------------------------------------|
  |`Geometry::Geometry`         |_None_         |Input Geometry (may include other than curves)                    |
  |`Surface::Geometry`          |_None_         |Surface geometry to attach hair curves to                         |
  |`Surface::Object`            |None           |Surface Object to attach to (needs to have matching transforms)   |
  |`Surface UV Map::Vector`     |[0.0, 0.0, 0.0]|Surface UV map used for attachment                                |
  |`Surface Rest Position::Bool`|False          |Set the surface mesh into its rest position before attachment     |
  |`Sample Attachment UV::Bool` |True           |Sample the surface UV map at the attachment point                 |
  |`Snap to Surface::Bool`      |True           |Snap the root of each curve to the closest surface point          |
  |`Align to Surface Normal::Bool`|True           |Align the curve to the surface normal (needs a guide as reference)|
  |`Blend along Curve::Float`   |0.0            |Blend deformation along each curve from the root                  |
  
  : {tbl-colwidths="[10, 15, 80]"}
  
  
  
  '''
# ---
